core/factorization/ilu.cpp

```diff
@@ -79,16 +93,36 @@ std::unique_ptr<Composition<ValueType>> Ilu<ValueType, IndexType>::generate_l_u(
         local_system_matrix.get(), false));
 
     // Compute LU factorization
-    exec->run(ilu_factorization::make_compute_ilu(local_system_matrix.get()));
-
+    if (std::dynamic_pointer_cast<const OmpExecutor>(exec) ||
+        parameters_.algorithm == factorize_algorithm::syncfree) {
+        auto sparsity =
+            share(gko::matrix::SparsityCsr<ValueType, IndexType>::create_const(
+                exec, local_system_matrix->get_size(),
+                make_const_array_view(
+                    exec, local_system_matrix->get_num_stored_elements(),
+                    local_system_matrix->get_const_col_idxs()),
+                make_const_array_view(
+                    exec, local_system_matrix->get_size()[0] + 1,
+                    local_system_matrix->get_const_row_ptrs())));
+        ilu =
+            gko::experimental::factorization::Lu<ValueType, IndexType>::build()
+                .with_checked_lookup(true)
+                .with_symbolic_factorization(sparsity)
+                .on(exec)
+                ->generate(local_system_matrix)
+                ->get_combined();
```

Comment:
LU always generates the combined matrix first.
We can call `unpack` or rely on the original code here. I do not check there's any difference between the implementation.

Reply:
Right, LU creates combined matrix, but my question is, why call `get_combined` directly (which returns a sparse matrix, right?) when we could store the `Factorization` and get the L and U factors from that? I guess there's no reason to expect the performance would be different between the `Factorization` method for splitting the factors from a combined matrix vs the kernels called for turning the cuSPARSE output into factors?

Reply:
That's right. I did it because it reuse the following part and make the condition block smaller.
I can also use early return to avoid a large condition block actually.
I am not big fan of large condition block

Reply:
To enable this code path, I also add strategy argument to unpack function.